### DISCUSSION TITLE
Save eye tracking and video data periodically in multiple files

### DISCUSSION THREAD

**Discussion by derekalbert** [jspsych/jsPsych#3222] [Q&A] [answered]
I am programming an experiment for local (in-lab) deployment where I want to record both eye movements (using Webgazer) and video of the participant. Both eye-tracking and video data are locked to ~6.5 second trials, of which there is a total of 234. 

I have implemented both extensions correctly, but the output CSV is very large. Of course, there are concerns about timing accuracy. I am also concerned about the browser crashing while writing the file or exceeding the browser's cache limit (I presume there is one). Furthermore, I am concerned about working with these very large files later to extract the eye tracking and, especially, video data. 

I am wondering if there is a way to save the eye-tracking and video data separately, progressively throughout the experiment. Perhaps this could be done in blocks, or even per trial. Presumably this would also involve clearing the old eye-tracking and video data from memory. Perhaps it would be easier to save everything (including task data) and clearing all of it in chunks?

Would greatly appreciate the community's help!

**Comment by Shaobin-Jiang** [comment]
You can set up a local server and send whatever you might want to save when you when to save it to the server.

**Reply by derekalbert** [reply, marked as answer]
Thanks Shaobin,

The solution I ended up with was to include this timeline item at the end of each block:

        timeline.push({
            type: jsPsychCallFunction,
            func: function () {
                var dataToSave = jsPsych.data.get().csv();
                saveData(dataToSave, SONAID, task, blockType, currentBlockNumber);
                jsPsych.data.reset();
            }
        });

    function saveData(data, sona, task, type, block) {
        var xhr = new XMLHttpRequest();
        xhr.open('POST', 'php/write_data.php'); // 'write_data.php' is the path to the php file described above.
        xhr.setRequestHeader('Content-Type', 'application/json');
        xhr.send(JSON.stringify({ filedata: data, sona: sona, task: task, type: type, block: block }));
    }

As you suggested, I used PHP on my local server to write the CSV files whenever my saveData function was called.
